library(dplyr)
library(MOVICS)
library(ggplot2)
library(survival)
library(survminer)
library(readr)
library(clusterProfiler)
library(org.Hs.eg.db)  
library(enrichplot)    
library(DOSE)
library(corrplot)
library(scRNAtoolVis)
library(GseaVis)
#è¯»å–MOVICSåˆ†å‹ç»“æœ
load("~/NSCLC_subtyping/Unsupervised Clustering/MULTI-OMICS/outputdata/cmoic.cluster.results.rda")

#è¯»å–ä¸´åºŠä¸ç”Ÿå­˜æ•°æ®
load("~/NSCLC_subtyping/INTERGRATION AND ANALYSIS/outputdata/commonsample.clinical.rda")
load("~/NSCLC_subtyping/INTERGRATION AND ANALYSIS/outputdata/commonsample.survival.rda")


# ç¤ºä¾‹æ•°æ®:TCGAçš„ä¹³è…ºç™Œæ•°æ®
load(system.file("extdata", "brca.tcga.RData", package = "MOVICS", mustWork = TRUE))
load(system.file("extdata", "brca.yau.RData",  package = "MOVICS", mustWork = TRUE))

#ç”Ÿå­˜åˆ†æ -------------------------------------------------------------------
setwd("~/NSCLC_subtyping/Downstream Analysis of MOVICS clusters/survival_output")
#Coxæ¯”ä¾‹é£é™©å›å½’æ¨¡å‹
#æ•°æ®å‡†å¤‡
#
all(rownames(cmoic.nsclc) == rownames(commonsample.survival ))  
surv.info <- data.frame(
  OS.time  = as.numeric(commonsample.survival$OS.time),  # ç”Ÿå­˜æ—¶é—´ï¼ˆå¤©ï¼‰
  OS.stat   = as.numeric(commonsample.survival$OS),       # ç”Ÿå­˜çŠ¶æ€ï¼ˆ0/1ï¼‰
  PFI.time = as.numeric(commonsample.survival$PFI.time),
  PFI.stat =as.numeric(commonsample.survival$PFI),
  cluster = cmoic.nsclc$clust.res$clust,                       # åˆ†å­äºšå‹ï¼ˆCS1-CS5ï¼‰
  stage   = commonsample.clinical$ajcc_pathologic_stage.diagnoses,  # ä¸´åºŠåˆ†æœŸï¼ˆStage I-IVï¼‰
  row.names = rownames(commonsample.survival)
)
surv.info$OS.time_month <- surv.info$OS.time / 30.4368  # ä½¿ç”¨365/12â‰ˆ30.4368å¤©/æœˆçš„ç²¾ç¡®è½¬æ¢
surv.info$PFI.time_month <- surv.info$PFI.time / 30.4368  # ä½¿ç”¨365/12â‰ˆ30.4368å¤©/æœˆçš„ç²¾ç¡®è½¬æ¢
all(rownames(cmoic.nsclc) == rownames(surv.info ))  #æ£€æŸ¥è¡Œé¡ºåº
#å‘½åç±»åˆ«
surv.info$cluster <- gsub("1", "CS1", surv.info$cluster)
surv.info$cluster <- gsub("2", "CS2", surv.info$cluster)
surv.info$cluster <- gsub("3",  "CS3", surv.info$cluster)
surv.info$cluster <- gsub("4", "CS4", surv.info$cluster)
surv.info$cluster <- gsub("5", "CS5", surv.info$cluster)
# å°†å­åˆ†æœŸåˆå¹¶ä¸ºä¸»åˆ†æœŸ
# # åˆ é™¤æ‰€æœ‰ "Stage " å‰ç¼€ï¼ˆåŒ…æ‹¬ç©ºæ ¼ï¼‰
surv.info$stage <- gsub("^Stage\\s*", "", surv.info$stage, ignore.case = TRUE)
surv.info$stage <- gsub("Stage I", "I", surv.info$stage)
surv.info$stage <- gsub("IA", "I", surv.info$stage)
surv.info$stage <- gsub("IB", "I", surv.info$stage)
surv.info$stage <- gsub("II", "II", surv.info$stage)
surv.info$stage <- gsub("IIA", "II", surv.info$stage)
surv.info$stage <- gsub("IIB", "II", surv.info$stage)
surv.info$stage <- gsub("III", "III", surv.info$stage)
surv.info$stage <- gsub("IIIA", "III", surv.info$stage)
surv.info$stage <- gsub("IIIB", "III", surv.info$stage)
surv.info$stage <- gsub("IV", "IV", surv.info$stage)
surv.info[] <- lapply(surv.info, function(col) {
  col[trimws(col) == ""] <- NA
  return(col)
})   


# ä½¿ç”¨survivalåŒ…ä¸­çš„survfitå‡½æ•°è®¡ç®—kaplan-Meierç”Ÿå­˜ä¼°è®¡
OS.fit <- survfit(Surv(OS.time_month, OS.stat) ~ cluster, data = surv.info)
PFI.fit <- survfit(Surv(PFI.time_month, PFI.stat) ~ cluster, data = surv.info)
#æŸ¥çœ‹å®Œæ•´çš„ç”Ÿå­˜è¡¨æ ¼
summary(OS.fit)$table
summary(PFI.fit)$table
# records n.max n.start events     rmean se(rmean)   median  0.95LCL  0.95UCL
# cluster=1     114   114     114     43  93.32458  16.98745 62.81869 37.29039       NA
# cluster=2     112   112     112     33 122.29048  15.74697 61.60306 49.38101       NA
# cluster=3     137   137     137     51  92.06291  16.26252 45.27414 28.41954       NA
# cluster=4     121   121     121     59  73.01746  15.41266 26.57967 18.66162 41.33155
# cluster=5      56    56      56     22 107.65940  21.37707 47.63970 29.01093       NA
# ç»˜åˆ¶ç”Ÿå­˜æ›²çº¿
#OS
pdf(file = "Kaplan-Meier Analysis of Overall Survival by CS Subtype.pdf", 
    width = 8,        # å®½åº¦ï¼ˆè‹±å¯¸ï¼‰
    height = 8 )       # é«˜åº¦ï¼ˆè‹±å¯¸ï¼‰
ggsurvplot(OS.fit, 
           surv.median.line = "hv",           # åŒæ—¶æ˜¾ç¤ºä¸­ä½ç”Ÿå­˜æ—¶é—´çš„å‚ç›´å’Œæ°´å¹³å‚è€ƒçº¿
           pval = T,# æ˜¾ç¤ºç»Ÿè®¡æ£€éªŒpå€¼
           conf.int = F,                      # æ˜¾ç¤ºç½®ä¿¡åŒºé—´
           risk.table = T,                    # æ˜¾ç¤ºé£é™©è¡¨æ ¼ï¼Œå±•ç¤ºæ¯ä¸ªæ—¶é—´ç‚¹çš„é£é™©æ•°é‡
           risk.table.col = "strata",         # é£é™©è¡¨æ ¼æŒ‰ç…§æ€§åˆ«ç€è‰²
           xlab = "Time(months)",               # xè½´æ ‡ç­¾
           ylab = "Overall Survival probability",     # yè½´æ ‡ç­¾
           legend.title = "Subtype", # å›¾ä¾‹æ ‡é¢˜
           legend.labs = c("CS1", "CS2","CS3","CS4","CS5"), # å›¾ä¾‹æ ‡ç­¾
           ggtheme = theme_minimal() +  # ä½¿ç”¨æç®€ä¸»é¢˜
             theme(
               # è°ƒæ•´å›¾ä¾‹æ–‡å­—
               legend.text = element_text(size = 14, color = "black"),  # å›¾ä¾‹é¡¹ï¼ˆCS1-CS5ï¼‰å­—ä½“å¤§å°
               legend.title = element_text(size = 16, face = "bold"),   # å›¾ä¾‹æ ‡é¢˜ï¼ˆSubtypeï¼‰å­—ä½“å¤§å°
               # åæ ‡è½´åˆ»åº¦æ ‡ç­¾å­—ä½“ï¼ˆæ•°å­—éƒ¨åˆ†ï¼‰
               axis.text.x = element_text(size = 12, color = "black"),  # Xè½´åˆ»åº¦æ–‡å­—
               axis.text.y = element_text(size = 12, color = "black"),  # Yè½´åˆ»åº¦æ–‡å­—
               # åæ ‡è½´æ ‡é¢˜å­—ä½“ï¼ˆ"OS Time(months)"å’Œ"Survival Probability"ï¼‰
               axis.title.x = element_text(size = 14, face = "bold", margin = margin(t=10)),  # Xè½´æ ‡é¢˜
               axis.title.y = element_text(size = 14, face = "bold", margin = margin(r=10)),  # Yè½´æ ‡é¢˜
               axis.line = element_line(color = "black", size = 0.8),  # è®¾ç½®è½´çº¿ä¸ºé»‘è‰²ç²—çº¿
               panel.grid.major = element_blank(),  # ç§»é™¤ä¸»ç½‘æ ¼çº¿
               panel.grid.minor = element_blank()   # ç§»é™¤æ¬¡ç½‘æ ¼çº¿
             ),
           break.x.by = 12,                  # xè½´åˆ»åº¦é—´éš”
           palette = c("#2EC4B6", "#E71D36", "#FF9F1C", "#BDD5EA", "#FFA5BA")) # è‡ªå®šä¹‰é¢œè‰²
dev.off()
#PFI
pdf(file = "Kaplan-Meier Analysis of Progression Free Interval by CS Subtype.pdf", 
    width = 8,        # å®½åº¦ï¼ˆè‹±å¯¸ï¼‰
    height = 8 )       # é«˜åº¦ï¼ˆè‹±å¯¸ï¼‰
ggsurvplot(PFI.fit, 
           surv.median.line = "hv",           # åŒæ—¶æ˜¾ç¤ºä¸­ä½ç”Ÿå­˜æ—¶é—´çš„å‚ç›´å’Œæ°´å¹³å‚è€ƒçº¿
           pval = T,                          # æ˜¾ç¤ºç»Ÿè®¡æ£€éªŒpå€¼
           conf.int = F,                      # ä¸æ˜¾ç¤ºç½®ä¿¡åŒºé—´
           risk.table = T,                    # æ˜¾ç¤ºé£é™©è¡¨æ ¼ï¼Œå±•ç¤ºæ¯ä¸ªæ—¶é—´ç‚¹çš„é£é™©æ•°é‡
           risk.table.col = "strata",         # é£é™©è¡¨æ ¼æŒ‰ç…§æ€§åˆ«ç€è‰²
           xlab = "Time(months)",               # xè½´æ ‡ç­¾
           ylab = "Progression Free Survival probability",     # yè½´æ ‡ç­¾
           legend.title = "Subtype", # å›¾ä¾‹æ ‡é¢˜
           legend.labs = c("CS1", "CS2","CS3","CS4","CS5"), # å›¾ä¾‹æ ‡ç­¾
           ggtheme = theme_minimal() +  # ä½¿ç”¨æç®€ä¸»é¢˜
             theme(
               # è°ƒæ•´å›¾ä¾‹æ–‡å­—
               legend.text = element_text(size = 14, color = "black"),  # å›¾ä¾‹é¡¹ï¼ˆCS1-CS5ï¼‰å­—ä½“å¤§å°
               legend.title = element_text(size = 16, face = "bold"),   # å›¾ä¾‹æ ‡é¢˜ï¼ˆSubtypeï¼‰å­—ä½“å¤§å°
               # åæ ‡è½´åˆ»åº¦æ ‡ç­¾å­—ä½“ï¼ˆæ•°å­—éƒ¨åˆ†ï¼‰
               axis.text.x = element_text(size = 12, color = "black"),  # Xè½´åˆ»åº¦æ–‡å­—
               axis.text.y = element_text(size = 12, color = "black"),  # Yè½´åˆ»åº¦æ–‡å­—
               # åæ ‡è½´æ ‡é¢˜å­—ä½“
               axis.title.x = element_text(size = 14, face = "bold", margin = margin(t=10)),  # Xè½´æ ‡é¢˜
               axis.title.y = element_text(size = 14, face = "bold", margin = margin(r=10)),  # Yè½´æ ‡é¢˜
               axis.line = element_line(color = "black", size = 0.8),  # è®¾ç½®è½´çº¿ä¸ºé»‘è‰²ç²—çº¿
               panel.grid.major = element_blank(),  # ç§»é™¤ä¸»ç½‘æ ¼çº¿
               panel.grid.minor = element_blank()   # ç§»é™¤æ¬¡ç½‘æ ¼çº¿
             ),
           break.x.by = 12,                  # xè½´åˆ»åº¦é—´éš”
           palette = c("#2EC4B6", "#E71D36", "#FF9F1C", "#BDD5EA", "#FFA5BA")) # è‡ªå®šä¹‰é¢œè‰²
dev.off()


#K-Mç”Ÿå­˜æ›²çº¿
# surv.data <-  data.frame(
#   futime
#   = as.numeric(commonsample.survival$OS.time),  # ç”Ÿå­˜æ—¶é—´ï¼ˆå¤©ï¼‰
#   fustat
#   = as.numeric(commonsample.survival$OS),       # ç”Ÿå­˜çŠ¶æ€ï¼ˆ0/1ï¼‰
#   row.names = rownames(commonsample.survival)
# )
# 
# surv.nsclc <- compSurv(moic.res         = cmoic.nsclc,
#                        surv.info        = surv.info,
#                        convt.time       = "m", # æŠŠå¤©å˜æˆæœˆ
#                        surv.median.line = "h",
#                        xyrs.est         = c(5,10), # è®¡ç®—5å¹´å’Œ10å¹´ç”Ÿå­˜ç‡
#                        fig.name         = "KAPLAN-MEIER CURVE OF CONSENSUSMOIC")




#Coxæ¯”ä¾‹é£é™©å›å½’æ¨¡å‹
#æ£€æŸ¥æ•°æ®å®Œæ•´æ€§
str(surv.info)
sum(is.na(surv.info))  # å¤„ç†ç¼ºå¤±å€¼ï¼ˆå¦‚åˆ é™¤æˆ–å¡«è¡¥ï¼‰
#[1] 50
surv.info.cox <- surv.info
surv.info.cox <- na.omit(surv.info.cox) #å»é™¤äº†æ²¡æœ‰ç”Ÿå­˜æ•°æ®å’Œåˆ†æœŸæ•°æ®çš„æ ·æœ¬ï¼Œå‰©ä½™535ä¸ª


#è½¬æ¢ä¸ºå› å­
surv.info.cox$cluster <- factor(surv.info.cox$cluster, levels = c("CS1", "CS2", "CS3", "CS4","CS5"))
surv.info.cox$stage <- factor(surv.info.cox$stage, levels = c("I", "II", "III", "IV")) 
table(surv.info.cox$stage, useNA = "always")
# I   II  III   IV <NA> 
#   271  159   90   15    0 
table(surv.info.cox$cluster, useNA = "always")
# CS1  CS2  CS3  CS4  CS5 <NA> 
#   112  112  136  119   56    0
# æ‹Ÿåˆå¤šå˜é‡Coxæ¨¡å‹
OS.cox.ph <- coxph(
  Surv(time = OS.time, event = OS.stat) ~ cluster + stage,
  data = surv.info.cox
)
PFI.cox.ph <- coxph(
  Surv(time = PFI.time, event = PFI.stat) ~ cluster + stage,
  data = surv.info.cox
)
# æŸ¥çœ‹æ¨¡å‹æ‘˜è¦
summary(OS.cox.ph)
# Call:
#   coxph(formula = Surv(time = OS.time, event = OS.stat) ~ cluster + 
#           stage, data = surv.info.cox)
# 
# n= 535, number of events= 213 
# 
# coef exp(coef) se(coef)      z Pr(>|z|)    
# clusterCS2 -0.04662   0.95445  0.21432 -0.218  0.82780    
# clusterCS3  0.08210   1.08556  0.21358  0.384  0.70069    
# clusterCS4  0.28931   1.33551  0.20604  1.404  0.16028    
# clusterCS5  0.11703   1.12416  0.24569  0.476  0.63383    
# stageII     0.51386   1.67173  0.16808  3.057  0.00223 ** 
#   stageIII    0.85265   2.34586  0.17925  4.757 1.97e-06 ***
#   stageIV     1.37790   3.96656  0.31538  4.369 1.25e-05 ***
#   ---
#   Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1
# 
# exp(coef) exp(-coef) lower .95 upper .95
# clusterCS2    0.9545     1.0477    0.6271     1.453
# clusterCS3    1.0856     0.9212    0.7143     1.650
# clusterCS4    1.3355     0.7488    0.8918     2.000
# clusterCS5    1.1242     0.8896    0.6945     1.820
# stageII       1.6717     0.5982    1.2025     2.324
# stageIII      2.3459     0.4263    1.6509     3.333
# stageIV       3.9666     0.2521    2.1378     7.360
# 
# Concordance= 0.628  (se = 0.024 )
# Likelihood ratio test= 35.72  on 7 df,   p=8e-06
# Wald test            = 38.6  on 7 df,   p=2e-06
# Score (logrank) test = 41.5  on 7 df,   p=6e-07
summary(PFI.cox.ph)
# Call:
#   coxph(formula = Surv(time = PFI.time, event = PFI.stat) ~ cluster + 
#           stage, data = surv.info.cox)
# 
# n= 535, number of events= 205 
# 
# coef exp(coef) se(coef)      z Pr(>|z|)    
# clusterCS2 -0.29098   0.74753  0.23457 -1.240 0.214792    
# clusterCS3  0.28209   1.32590  0.21195  1.331 0.183215    
# clusterCS4  0.47317   1.60508  0.20556  2.302 0.021345 *  
#   clusterCS5  0.02334   1.02361  0.26649  0.088 0.930210    
# stageII     0.64186   1.90000  0.16515  3.886 0.000102 ***
#   stageIII    0.72350   2.06165  0.18931  3.822 0.000132 ***
#   stageIV     1.43231   4.18836  0.34077  4.203 2.63e-05 ***
#   ---
#   Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1
# 
# exp(coef) exp(-coef) lower .95 upper .95
# clusterCS2    0.7475     1.3377    0.4720     1.184
# clusterCS3    1.3259     0.7542    0.8752     2.009
# clusterCS4    1.6051     0.6230    1.0728     2.401
# clusterCS5    1.0236     0.9769    0.6072     1.726
# stageII       1.9000     0.5263    1.3746     2.626
# stageIII      2.0616     0.4850    1.4226     2.988
# stageIV       4.1884     0.2388    2.1477     8.168
# 
# Concordance= 0.654  (se = 0.021 )
# Likelihood ratio test= 42.2  on 7 df,   p=5e-07
# Wald test            = 44.21  on 7 df,   p=2e-07
# Score (logrank) test = 47.24  on 7 df,   p=5e-08

library(survminer)
# åŸºç¡€æ£®æ—å›¾
pdf("OS.forest plot.pdf", width = 10, height = 8)
ggforest(
  model = OS.cox.ph,           # Coxæ¨¡å‹å¯¹è±¡
  data = surv.info,         # ä½¿ç”¨çš„æ•°æ®
  main = "Hazard Ratio (95% CI) for Cluster and Stage",
  fontsize = 1.0,          # æ–‡å­—å°ºå¯¸
  noDigits = 3             # å°æ•°ä½æ•°
)
dev.off()

pdf("PFI.forest plot.pdf", width = 10, height = 8)
ggforest(
  model = PFI.cox.ph,           # Coxæ¨¡å‹å¯¹è±¡
  data = surv.info,         # ä½¿ç”¨çš„æ•°æ®
  main = "Hazard Ratio (95% CI) for Cluster and Stage",
  fontsize = 1.0,          # æ–‡å­—å°ºå¯¸
  noDigits = 3             # å°æ•°ä½æ•°
)
dev.off()


# æ¯”è¾ƒä¸åŒäºšå‹é—´çªå˜é¢‘ç‡ -------------------------------------------------------------
setwd("~/NSCLC_subtyping/Downstream Analysis of MOVICS clusters/Mutation_output")
#æå–ä¸´åºŠä¿¡æ¯ä½œä¸ºannotation
annCol    <- data.frame(project_id=commonsample.clinical$project_id.project,
                        gender=commonsample.clinical$gender.demographic,
                        stage=commonsample.clinical$ajcc_pathologic_stage.diagnoses,
                        row.names = rownames(commonsample.clinical)
)
#stageæ•°æ®ç»Ÿä¸€æ ¼å¼
annCol$stage<- gsub("Stage IA", "Stage I", annCol$stage)
annCol$stage <- gsub("Stage IB", "Stage I", annCol$stage)
annCol$stage <- gsub("Stage IIA", "Stage II", annCol$stage)
annCol$stage <- gsub("Stage IIB", "Stage II", annCol$stage)
annCol$stage <- gsub("Stage IIIA", "Stage III", annCol$stage)
annCol$stage <- gsub("Stage IIIB", "Stage III", annCol$stage)
annCol[] <- lapply(annCol, function(col) {
  col[trimws(col) == ""] <- "unknown"
  return(col)
})

#ä¸ºannotationè®¾ç½®é¢œè‰²
annColors <- list( gender =c("male" = "#2CA02C", "female" = "#D62728"),
                   project_id=c("TCGA-LUAD" =  "#D4AF37", "TCGA-LUSC" = "#6C5B7B"),
                   stage = c("Stage I"   = "#8DA0CB", 
                                        "Stage II"  = "#E78AC3",
                                        "Stage III" = "#FC8D62",  #
                                        "Stage IV"  = "#66C2A5" ,
                                        "unknown" = "white"))

mut.brca <- compMut(moic.res     = cmoic.nsclc,
                    mut.matrix   = multiomics.data$MUTA, # 0/1çŸ©é˜µ
                    test.method = "fisher",
                    p.adj.method = "BH",
                    doWord       = TRUE, # ç”ŸæˆWordæ–‡æ¡£
                    doPlot       = TRUE, # draw OncoPrint
                    freq.cutoff  = 0.05, # ä¿ç•™åœ¨è‡³å°‘5%çš„æ ·æœ¬ä¸­çªå˜çš„åŸºå› 
                    p.adj.cutoff = 0.05, # ä¿ç•™padj<0.05çš„åŸºå› 
                    innerclust   = TRUE, # åœ¨æ¯ä¸ªäºšå‹ä¸­è¿›è¡Œèšç±»
                    annCol       = annCol, 
                    annColors    = annColors, 
                    clust.col = c("#2EC4B6", "#E71D36", "#FF9F1C", "#BDD5EA", "#FFA5BA"),
                    width        = 8, 
                    height       = 6,
                    fig.name     = "ONCOPRINT FOR SIGNIFICANT MUTATIONS",
                    tab.name     = "INDEPENDENT TEST BETWEEN SUBTYPE AND MUTATION")

# è‚¿ç˜¤çªå˜è´Ÿè·åˆ†æ ----------------------------------------------------------------
# if (!requireNamespace("BiocManager", quietly=TRUE))
#   install.packages("BiocManager")
# BiocManager::install("BioinformaticsFMRP/TCGAbiolinksGUI.data",lib = "~/R/x86_64-pc-linux-gnu-library/4.4")
# BiocManager::install("TCGAbiolinks",lib = "~/R/x86_64-pc-linux-gnu-library/4.4")
# install.packages("/home/yanbin/RåŒ…/TCGAbiolinks_2.34.1.tar.gz",
#                  repos = NULL,
#                  type = "source",
#                  lib = "/home/yanbin/R/x86_64-pc-linux-gnu-library/4.4")
                  
library(TCGAbiolinks)
#ä¸‹è½½mafæ–‡ä»¶å¹¶æ•´ç†
LUAD.maf <- GDCquery(
  project = "TCGA-LUAD", 
  data.category = "Simple Nucleotide Variation",
  data.type = "Masked Somatic Mutation",
  access = "open"
)
GDCdownload(LUAD.maf)
GDCprepare(LUAD.maf, save = T,save.filename = "TCGA-LUAD_SNP.Rdata")


LUSC.maf <- GDCquery(
  project = "TCGA-LUSC", 
  data.category = "Simple Nucleotide Variation",
  data.type = "Masked Somatic Mutation",
  access = "open"
)
GDCdownload(LUSC.maf)
GDCprepare(LUSC.maf, save = T,save.filename = "TCGA-LUSC_SNP.Rdata")

load("~/NSCLC_subtyping/Downstream Analysis of MOVICS clusters/TCGA-LUAD_SNP.Rdata")
maf.luad<- data
load("~/NSCLC_subtyping/Downstream Analysis of MOVICS clusters/TCGA-LUSC_SNP.Rdata")
maf.lusc<- data
maf_combine <- rbind(maf.luad,maf.lusc)
maf <- data.frame(Tumor_Sample_Barcode=maf_combine$Tumor_Sample_Barcode,
                  Hugo_Symbol=maf_combine$Hugo_Symbol,
                  Chromosome=maf_combine$Chromosome,
                  Start_Position=maf_combine$Start_Position,
                  End_Position=maf_combine$End_Position,
                  Variant_Classification=maf_combine$Variant_Classification,
                  Variant_Type=maf_combine$Variant_Type,
                  Reference_Allele=maf_combine$Reference_Allele,
                  Tumor_Seq_Allele1=maf_combine$Tumor_Seq_Allele1,
                  Tumor_Seq_Allele2=maf_combine$Tumor_Seq_Allele2)
maf$Tumor_Sample_Barcode <- substr(maf$Tumor_Sample_Barcode, start = 1, stop = 15)

tmb.nsclc <- compTMB(moic.res     = cmoic.nsclc,
                    maf          = maf,
                    rmDup        = TRUE, # remove duplicated variants per sample
                    rmFLAGS      = FALSE, # keep FLAGS mutations
                    exome.size   = 38, # estimated exome size
                    clust.col = c("#2EC4B6", "#E71D36", "#FF9F1C", "#BDD5EA", "#FFA5BA"),
                    test.method  = "nonparametric", # statistical testing method
                    fig.name     = "DISTRIBUTION OF TMB AND TITV",
                    width=6,
                    height =6)






#library(maftools)


# äºšå‹é—´å·®å¼‚åˆ†æ -----------------------------------------------------------------
#è¯»å–tpmæ•°æ®
LUAD_tpm<- read.table("/home/data/yanbin/NSCLC_Dataset/LUAD/GEXP/tpm/TCGA-LUAD.star_tpm.tsv", header = TRUE, sep = "\t",row.names=1)
LUSC_tpm <- read.table("/home/data/yanbin/NSCLC_Dataset/LUSC/GEXP/tpm/TCGA-LUSC.star_tpm.tsv", header = TRUE, sep = "\t",row.names=1)
#å¤„ç†æ ¼å¼
all(rownames(LUAD_tpm)==rownames(LUSC_tpm))#æ£€æŸ¥è¡Œåæ˜¯å¦ä¸€è‡´
nsclc_tpm <- cbind(LUAD_tpm,LUSC_tpm)
colnames(nsclc_tpm) <- substr(colnames(nsclc_tpm), start = 1, stop = 15)
colnames(nsclc_tpm) <- gsub("\\.", "-", colnames(nsclc_tpm))
nsclc_tpm <- nsclc_tpm[,rownames(commonsamples.clinical.merge)]
nsclc_tpm$Ensembl_ID <- rownames(nsclc_tpm)
gene_id <- read.table("~/NSCLC_subtyping/Downstream Analysis of MOVICS clusters/DESeq2 inputdata/gencode.v36.annotation.gtf.gene.probemap",header = TRUE, sep = "\t")
gene_id <- gene_id[ , c(1,2)]
nsclc_tpm <- merge(gene_id, nsclc_tpm, by.y  = "Ensembl_ID", by.x = "id" )#åŸºå› åè½¬æ¢
nsclc_tpm <- distinct(nsclc_tpm, gene, .keep_all = T)#å»é‡ï¼ˆé‡å¤çš„åªä¿ç•™ç¬¬ä¸€è¡Œï¼‰

# å¯ä»¥ç”¨limmaåŒ…ä¸­çš„averepså‡½æ•°ï¼Œæˆ–è€…ä¹Ÿå¯ä»¥ä½¿ç”¨aggregateå‡½æ•°å–å¹³å‡
# library(limma)
# nsclc_counts <- avereps(nsclc_counts, nsclc_counts$gene)

#æŠŠåŸºå› åè½¬æ¢ä¸ºè¡Œå
rownames(nsclc_tpm) <- nsclc_tpm$gene
nsclc_tpm <- nsclc_tpm[ , -c(1,2)]
dim(nsclc_tpm) # [1] 59427   551
head(nsclc_tpm)[1:5, 1:5]

#å·®å¼‚åˆ†ælimmaæ–¹æ³•
setwd( "~/NSCLC_subtyping/Downstream Analysis of MOVICS clusters/limma output")
runDEA(dea.method = "limma",
       expr       = nsclc_tpm, 
       moic.res   = cmoic.nsclc,
       prefix     = "TCGA-NSCLC")
# data('pbmc.markers')
library(corrplot)
library(scRNAtoolVis)
#è¯»å–å·®å¼‚åˆ†æç»“æœ
CS1_vs_others <- read.table("/home/yanbin/NSCLC_subtyping/Downstream Analysis of MOVICS clusters/limma output/consensusMOIC_TCGA-NSCLC_limma_test_result.CS1_vs_Others.txt",header=TRUE,sep="\t",row.names = 1)  
CS2_vs_others <- read.table("/home/yanbin/NSCLC_subtyping/Downstream Analysis of MOVICS clusters/limma output/consensusMOIC_TCGA-NSCLC_limma_test_result.CS2_vs_Others.txt",header=TRUE,sep="\t",row.names = 1)  
CS3_vs_others <- read.table("/home/yanbin/NSCLC_subtyping/Downstream Analysis of MOVICS clusters/limma output/consensusMOIC_TCGA-NSCLC_limma_test_result.CS3_vs_Others.txt",header=TRUE,sep="\t",row.names = 1)  
CS4_vs_others <- read.table("/home/yanbin/NSCLC_subtyping/Downstream Analysis of MOVICS clusters/limma output/consensusMOIC_TCGA-NSCLC_limma_test_result.CS4_vs_Others.txt",header=TRUE,sep="\t",row.names = 1)  
CS5_vs_others <- read.table("/home/yanbin/NSCLC_subtyping/Downstream Analysis of MOVICS clusters/limma output/consensusMOIC_TCGA-NSCLC_limma_test_result.CS5_vs_Others.txt",header=TRUE,sep="\t",row.names = 1)  
#æ•°æ®å‡†å¤‡
CS1_DEG <- data.frame(p_val = CS1_vs_others$pvalue,
                      avg_log2FC = CS1_vs_others$log2fc,
                      p_val_adj = CS1_vs_others$padj,
                      cluster = "CS1",
                      gene = rownames(CS1_vs_others),
                      row.names = rownames(CS1_vs_others)
)
CS1_DEG <- 
CS2_DEG <- data.frame(p_val = CS2_vs_others$pvalue,
                      avg_log2FC = CS2_vs_others$log2fc,
                      p_val_adj = CS2_vs_others$padj,
                      cluster = "CS2",
                      gene = rownames(CS2_vs_others),
                      row.names = rownames(CS2_vs_others)
)
CS3_DEG <- data.frame(p_val = CS3_vs_others$pvalue,
                      avg_log2FC = CS3_vs_others$log2fc,
                      p_val_adj = CS3_vs_others$padj,
                      cluster = "CS3",
                      gene = rownames(CS3_vs_others),
                      row.names = rownames(CS3_vs_others)
)
CS4_DEG <- data.frame(p_val = CS4_vs_others$pvalue,
                      avg_log2FC = CS4_vs_others$log2fc,
                      p_val_adj = CS4_vs_others$padj,
                      cluster = "CS4",
                      gene = rownames(CS4_vs_others),
                      row.names = rownames(CS4_vs_others)
)
CS5_DEG <- data.frame(p_val = CS5_vs_others$pvalue,
                      avg_log2FC = CS5_vs_others$log2fc,
                      p_val_adj = CS5_vs_others$padj,
                      cluster = "CS5",
                      gene = rownames(CS5_vs_others),
                      row.names = rownames(CS5_vs_others)
)
DEG_valcano <- rbind(CS5_DEG,CS4_DEG,CS3_DEG,CS2_DEG,CS1_DEG)
DEG_valcano_filtered <- subset(DEG_valcano, 
       p_val < 0.05 & abs(avg_log2FC) > 1.2)
#å¤šç»„ç«å±±å›¾
CSs_valcano <- jjVolcano(
  diffData = DEG_valcano,
  topGeneN = 10,
  log2FC.cutoff = 1.2,
  pvalue.cutoff= 0.05,
  adjustP.cutoff=0.01,
  col.type = "updown",
  aesCol = c('#0099CC','#CC3333'),
  tile.col = c("#2EC4B6", "#E71D36", "#FF9F1C", "#BDD5EA", "#FFA5AB"),
  cluster.order = rev(unique(DEG_valcano$cluster)),
  size  = 3.5,
  fontface = 'italic'
)
CSs_valcano
ggsave("volcano.png",width=16,height = 12,dpi=600)
ggsave("volcano.pdf",width=16,height = 12,dpi=600)


# è¯†åˆ«äºšå‹ç‰¹å®šç”Ÿç‰©æ ‡å¿—ç‰© -------------------------------------------------------------
# åŸºäºlimmaç»“æœè¯†åˆ«ä¸Šè°ƒçš„100ä¸ªåŸºå› 
# é€‰æ‹©ç”± log2FoldChange æ’åºçš„å·®å¼‚è¡¨è¾¾æœ€å¤šçš„åŸºå› ä½œä¸ºæ¯ä¸ªäºšå‹çš„ç”Ÿç‰©æ ‡å¿—ç‰©ï¼ˆé»˜è®¤æƒ…å†µä¸‹ï¼Œæ¯ä¸ªäºšå‹ 200 ä¸ªç”Ÿç‰©æ ‡å¿—ç‰©ï¼‰ã€‚è¿™äº›ç”Ÿç‰©æ ‡å¿—ç‰©åº”é€šè¿‡æ˜¾è‘—æ€§é˜ˆå€¼ï¼ˆä¾‹å¦‚ï¼Œæ ‡ç§°På€¼< 0.05 å¹¶è°ƒæ•´På€¼<0.05ï¼‰ï¼Œå¹¶ä¸”ä¸å¾—ä¸ä¸ºå…¶ä»–äºšå‹é‰´å®šçš„ä»»ä½•ç”Ÿç‰©æ ‡å¿—ç‰©é‡å ã€‚
setwd( "~/NSCLC_subtyping/Downstream Analysis of MOVICS clusters/biomaker output")
marker.up <- runMarker(moic.res      = cmoic.nsclc,
                       dea.method    = "limma", # name of DEA method
                       prefix        = "TCGA-NSCLC", # MUST be the same of argument in runDEA()
                       dat.path      = "~/NSCLC_subtyping/Downstream Analysis of MOVICS clusters/limma output", # path of DEA files
                       res.path      = getwd(), # path to save marker files
                       p.cutoff      = 0.05, # p cutoff to identify significant DEGs
                       p.adj.cutoff  = 0.05, # padj cutoff to identify significant DEGs
                       dirct         = "up", # direction of dysregulation in expression
                       n.marker      = 100, # number of biomarkers for each subtype
                       doplot        = TRUE, # generate diagonal heatmap
                       norm.expr     = nsclc_tpm, # use normalized expression as heatmap input
                       annCol        = annCol, # sample annotation in heatmap
                       annColors     = annColors, # colors for sample annotation
                       show_rownames = FALSE, # show no rownames (biomarker name)
                       fig.name      = "UPREGULATED BIOMARKER HEATMAP")

#åŸºäºlimmaç»“æœè¯†åˆ«ä¸‹è°ƒçš„100ä¸ªåŸºå› 
marker.dn <- runMarker(moic.res      = cmoic.nsclc,
                       dea.method    = "limma", # name of DEA method
                       prefix        = "TCGA-NSCLC", # MUST be the same of argument in runDEA()
                       dat.path      = "~/NSCLC_subtyping/Downstream Analysis of MOVICS clusters/limma output", # path of DEA files
                       res.path      = getwd(), # path to save marker files
                       p.cutoff      = 0.05, # p cutoff to identify significant DEGs
                       p.adj.cutoff  = 0.05, # padj cutoff to identify significant DEGs
                       dirct         = "down", # direction of dysregulation in expression
                       n.marker      = 100, # number of biomarkers for each subtype
                       doplot        = TRUE, # generate diagonal heatmap
                       norm.expr     = nsclc_tpm, # use normalized expression as heatmap input
                       annCol        = annCol, # sample annotation in heatmap
                       annColors     = annColors, # colors for sample annotation
                       show_rownames = FALSE, # show no rownames (biomarker name)
                       fig.name      = "DOWNREGULATED BIOMARKER HEATMAP")


# åŸºå› é›†å¯Œé›†åˆ†æç¡®å®šäºšå‹ç‰¹å¼‚æ€§åŠŸèƒ½é€šè·¯ GSEA gene set enrichment analysis-----------------------------------------------------------------
# if (!require("BiocManager")) install.packages("BiocManager")
# BiocManager::install(c("clusterProfiler", "org.Hs.eg.db", "enrichplot", "DOSE"))
library(clusterProfiler)
library(org.Hs.eg.db)  # äººç±»åŸºå› æ³¨é‡Šåº“
library(enrichplot)    # å¯è§†åŒ–
library(DOSE)          # æ”¯æŒå¯Œé›†åˆ†æ
library(msigdbr)     #åˆ†å­ç‰¹å¾æ•°æ®åº“çš„åŸºå› é›†
setwd( "~/NSCLC_subtyping/Downstream Analysis of MOVICS clusters/GSEA output")
## ç‰©ç§è®¾ç½®
organism = 'hsa'    #  äººç±»'hsa'   
OrgDb = 'org.Hs.eg.db'#äººç±»"org.Hs.eg.db"

#1ã€CS1äºšå‹å¯Œé›†åˆ†æ
#æå–äºšå‹1 vs å…¶ä»–äºšå‹çš„log2FCæ’åºåŸºå› åˆ—è¡¨
CS1_DEG_trans <- CS1_vs_others[,2,drop=FALSE]
CS1_DEG_trans$SYMBOL <- rownames(CS1_DEG_trans)

##åˆ›å»ºgseaåˆ†æçš„geneListï¼ˆåŒ…å«ä»å¤§åˆ°å°æ’åˆ—çš„log2FoldChangeå’ŒENTREZIDä¿¡æ¯ï¼‰
#è½¬åŒ–id  
SYMBOL_ENTERZID <- bitr(rownames(CS1_DEG_trans), 
           fromType = "SYMBOL",
           toType =  "ENTREZID",
           OrgDb = OrgDb) 
CS1_DEG_trans <- merge(CS1_DEG_trans, SYMBOL_ENTERZID , by='SYMBOL') #æŒ‰ç…§SYMBOLåˆå¹¶æ³¨é‡Šä¿¡æ¯


CS1_geneList <- CS1_DEG_trans$log2fc
names(CS1_geneList) <- CS1_DEG_trans$ENTREZID
CS1_geneList <- sort(CS1_geneList, decreasing = T)  #ä»å¤§åˆ°å°æ’åº

##gseaå¯Œé›† 
CS1_KEGG_kk_entrez <- gseKEGG(geneList     = CS1_geneList,
                          organism     = organism, #äººhsa é¼ mmu
                          pvalueCutoff = 0.25)  #å®é™…ä¸ºpadjé˜ˆå€¼å¯è°ƒæ•´ 
CS1_KEGG_kk <- DOSE::setReadable(CS1_KEGG_kk_entrez, 
                             OrgDb=OrgDb,
                             keyType='ENTREZID')#è½¬åŒ–id             

CS1_GO_kk_entrez <- gseGO(geneList     = CS1_geneList,
                      ont          = "ALL",  # "BP"ã€"MF"å’Œ"CC"æˆ–"ALL"
                      OrgDb        = OrgDb,
                      keyType      = "ENTREZID",
                      pvalueCutoff = 0.25)   #å®é™…ä¸ºpadjé˜ˆå€¼å¯è°ƒæ•´
CS1_GO_kk <- DOSE::setReadable(CS1_GO_kk_entrez, 
                           OrgDb=OrgDb,
                           keyType='ENTREZID')#è½¬åŒ–id 

save(CS1_KEGG_kk_entrez, CS1_GO_kk_entrez, file = "CS1_GSEA_result.RData")

#KEGGç»“æœå¯è§†åŒ–
##é€‰å–å¯Œé›†ç»“æœ
CS1_kk_gse <- CS1_KEGG_kk
#CS1_kk_gse_entrez <- CS1_KEGG_kk_entrez
###æ¡ä»¶ç­›é€‰
#ä¸€èˆ¬è®¤ä¸º|NES|>1ï¼ŒNOM pvalue<0.05ï¼ŒFDRï¼ˆpadjï¼‰<0.25çš„é€šè·¯æ˜¯æ˜¾è‘—å¯Œé›†çš„
CS1_kk_gse_cut <- CS1_kk_gse[CS1_kk_gse$pvalue<0.05 & CS1_kk_gse$p.adjust<0.25 & abs(CS1_kk_gse$NES)>1]
# CS1_kk_gse_cut_down <- CS1_kk_gse_cut[CS1_kk_gse_cut$NES < 0,]
# CS1_kk_gse_cut_up <- CS1_kk_gse_cut[CS1_kk_gse_cut$NES > 0,]
#é€‰æ‹©å±•ç°NESå‰å‡ ä¸ªé€šè·¯ 
# CS1_down_gsea <- CS1_kk_gse_cut_down[tail(order(CS1_kk_gse_cut_down$NES,decreasing = T),10),]
# CS1_up_gsea <- CS1_kk_gse_cut_up[head(order(CS1_kk_gse_cut_up$NES,decreasing = T),10),]
CS1_diff_gsea <- CS1_kk_gse_cut[head(order(abs(CS1_kk_gse_cut$NES),decreasing = T),10),]
#### ç»å…¸çš„GSEAå›¾ 
# CS1_up_gsea$Description
# i=2
# CS1_gseap1 <- gseaplot2(CS1_kk_gse,
#                         CS1_up_gsea$ID[i],#å¯Œé›†çš„IDç¼–å·
#                     title = CS1_up_gsea$Description[i],#æ ‡é¢˜
#                     color = "red", #GSEAçº¿æ¡é¢œè‰²
#                     base_size = 20,#åŸºç¡€å­—ä½“å¤§å°
#                     rel_heights = c(1.5, 0.5, 1),#å‰¯å›¾çš„ç›¸å¯¹é«˜åº¦
#                     subplots = 1:3,   #è¦æ˜¾ç¤ºå“ªäº›å‰¯å›¾ å¦‚subplots=c(1,3) #åªè¦ç¬¬ä¸€å’Œç¬¬ä¸‰ä¸ªå›¾
#                     ES_geom = "line", #enrichment scoreç”¨çº¿è¿˜æ˜¯ç”¨ç‚¹"dot"
#                     pvalue_table = T) #æ˜¾ç¤ºpvalueç­‰ä¿¡æ¯
# ggsave(CS1_gseap1, filename = 'CS1_GSEA_up_1.pdf', width =10, height =8)
#### åˆå¹¶ GSEAé€šè·¯ 
CS1_gseap <- gseaplot2(CS1_kk_gse,
                        CS1_diff_gsea$ID,#å¯Œé›†çš„IDç¼–å·
                    title = "CS1 KEGG Pathway",#æ ‡é¢˜
                    color = "red",#GSEAçº¿æ¡é¢œè‰²
                    base_size = 20,#åŸºç¡€å­—ä½“å¤§å°
                    rel_heights = c(1.5, 0.5, 1),#å‰¯å›¾çš„ç›¸å¯¹é«˜åº¦
                    subplots = 1:3, #è¦æ˜¾ç¤ºå“ªäº›å‰¯å›¾ å¦‚subplots=c(1,3) #åªè¦ç¬¬ä¸€å’Œç¬¬ä¸‰ä¸ªå›¾
                    ES_geom = "line",#enrichment scoreç”¨çº¿è¿˜æ˜¯ç”¨ç‚¹"dot"
                    pvalue_table = F) #æ˜¾ç¤ºpvalueç­‰ä¿¡æ¯
ggsave(CS1_gseap, filename = "CS1_GSEA_KEGG.png",width =16,height =16,dpi = 600 )

#GOç»“æœå¯è§†åŒ–
# ç»˜åˆ¶ç‚¹å›¾ï¼ˆæŒ‰p.adjustæ’åºï¼Œåˆ†é¢æ˜¾ç¤ºGOç±»åˆ«ï¼‰
dotplot(CS1_GO_kk,
        x = "NES",           # Xè½´ä¸ºNESå€¼ï¼ˆåæ˜ å¯Œé›†æ–¹å‘ï¼‰
        color = "p.adjust",  # é¢œè‰²è¡¨ç¤ºp.adjust
        showCategory = 10,   # æ¯ç±»æ˜¾ç¤ºTop10é€šè·¯
        split = "ONTOLOGY",  # æŒ‰GOç±»åˆ«åˆ†é¢
        font.size = 8) +  
  facet_grid(ONTOLOGY ~ ., scales = "free_y", space = "free") +
  ggtitle("CS1 GO Enrichment Analysis") +
  scale_color_gradient(low = "red", high = "blue")+
  theme(plot.title = element_text(hjust = 0.5)) # è‡ªå®šä¹‰é¢œè‰²
ggsave("CS1_GO_Dotplot.png", width = 8, height = 12,dpi = 600) 


#2ã€CS2äºšå‹å¯Œé›†åˆ†æ
#æå–äºšå‹2 vs å…¶ä»–äºšå‹çš„log2FCæ’åºåŸºå› åˆ—è¡¨
CS2_vs_others <- read.table("/home/yanbin/NSCLC_subtyping/Downstream Analysis of MOVICS clusters/limma output/consensusMOIC_TCGA-NSCLC_limma_test_result.CS2_vs_Others.txt",header=TRUE,sep="\t",row.names = 1)  
CS2_DEG_trans <- CS2_vs_others[,2,drop=FALSE]
CS2_DEG_trans$SYMBOL <- rownames(CS2_DEG_trans)

##åˆ›å»ºgseaåˆ†æçš„geneListï¼ˆåŒ…å«ä»å¤§åˆ°å°æ’åˆ—çš„log2FoldChangeå’ŒENTREZIDä¿¡æ¯ï¼‰
CS2_DEG_trans <- merge(CS2_DEG_trans, SYMBOL_ENTERZID , by='SYMBOL') #æŒ‰ç…§SYMBOLåˆå¹¶æ³¨é‡Šä¿¡æ¯
CS2_geneList <- CS2_DEG_trans$log2fc
names(CS2_geneList) <- CS2_DEG_trans$ENTREZID
CS2_geneList <- sort(CS2_geneList, decreasing = T)  #ä»å¤§åˆ°å°æ’åº

##gseaå¯Œé›† 
CS2_KEGG_kk_entrez <- gseKEGG(geneList     = CS2_geneList,
                              organism     = organism, #äººhsa é¼ mmu
                              pvalueCutoff = 0.25)  #å®é™…ä¸ºpadjé˜ˆå€¼å¯è°ƒæ•´ 
CS2_KEGG_kk <- DOSE::setReadable(CS2_KEGG_kk_entrez, 
                                 OrgDb=OrgDb,
                                 keyType='ENTREZID')#è½¬åŒ–id             

CS2_GO_kk_entrez <- gseGO(geneList     = CS2_geneList,
                          ont          = "ALL",  # "BP"ã€"MF"å’Œ"CC"æˆ–"ALL"
                          OrgDb        = OrgDb,
                          keyType      = "ENTREZID",
                          pvalueCutoff = 0.25)   #å®é™…ä¸ºpadjé˜ˆå€¼å¯è°ƒæ•´
CS2_GO_kk <- DOSE::setReadable(CS2_GO_kk_entrez, 
                               OrgDb=OrgDb,
                               keyType='ENTREZID')#è½¬åŒ–id 

save(CS2_KEGG_kk_entrez, CS2_GO_kk_entrez, file = "CS2_GSEA_result.RData")

#KEGGç»“æœå¯è§†åŒ–
##é€‰å–å¯Œé›†ç»“æœ
CS2_kk_gse <- CS2_KEGG_kk
#CS2_kk_gse_entrez <- CS2_KEGG_kk_entrez
###æ¡ä»¶ç­›é€‰
CS2_kk_gse_cut <- CS2_kk_gse[CS2_kk_gse$pvalue<0.05 & CS2_kk_gse$p.adjust<0.25 & abs(CS2_kk_gse$NES)>1]


#é€‰æ‹©å±•ç°NESå‰å‡ ä¸ªé€šè·¯ 
CS2_diff_gsea <- CS2_kk_gse_cut[head(order(abs(CS2_kk_gse_cut$NES),decreasing = T),10),]

#### åˆå¹¶ GSEAé€šè·¯ 
CS2_gseap <- gseaplot2(CS2_kk_gse,
                       CS2_diff_gsea$ID,#å¯Œé›†çš„IDç¼–å·
                       title = "CS2 KEGG Pathway",#æ ‡é¢˜
                       color = "red",#GSEAçº¿æ¡é¢œè‰²
                       base_size = 20,#åŸºç¡€å­—ä½“å¤§å°
                       rel_heights = c(1.5, 0.5, 1),#å‰¯å›¾çš„ç›¸å¯¹é«˜åº¦
                       subplots = 1:3, #è¦æ˜¾ç¤ºå“ªäº›å‰¯å›¾ å¦‚subplots=c(1,3) #åªè¦ç¬¬ä¸€å’Œç¬¬ä¸‰ä¸ªå›¾
                       ES_geom = "line",#enrichment scoreç”¨çº¿è¿˜æ˜¯ç”¨ç‚¹"dot"
                       pvalue_table = F) #æ˜¾ç¤ºpvalueç­‰ä¿¡æ¯
ggsave(CS2_gseap, filename = "CS2_GSEA_KEGG.png",width =16,height =16,dpi = 600)


#GOç»“æœå¯è§†åŒ–
# ç»˜åˆ¶ç‚¹å›¾ï¼ˆæŒ‰p.adjustæ’åºï¼Œåˆ†é¢æ˜¾ç¤ºGOç±»åˆ«ï¼‰
dotplot(CS2_GO_kk,
        x = "NES",           # Xè½´ä¸ºNESå€¼ï¼ˆåæ˜ å¯Œé›†æ–¹å‘ï¼‰
        color = "p.adjust",  # é¢œè‰²è¡¨ç¤ºp.adjust
        showCategory = 10,   # æ¯ç±»æ˜¾ç¤ºTop10é€šè·¯
        split = "ONTOLOGY",  # æŒ‰GOç±»åˆ«åˆ†é¢
        font.size = 8) +  
  facet_grid(ONTOLOGY ~ ., scales = "free_y", space = "free") +
  ggtitle("CS2 GO Enrichment Analysis") +
  scale_color_gradient(low = "red", high = "blue") +
  theme(plot.title = element_text(hjust = 0.5))  # è‡ªå®šä¹‰é¢œè‰²
ggsave("CS2_GO_Dotplot.png", width = 8, height = 12,,dpi = 600) 




#3ã€CS3äºšå‹å¯Œé›†åˆ†æ
#æå–äºš3 vs å…¶ä»–äºšå‹çš„log2FCæ’åºåŸºå› åˆ—è¡¨
CS3_DEG_trans <- CS3_vs_others[,2,drop=FALSE]
CS3_DEG_trans$SYMBOL <- rownames(CS3_DEG_trans)

##åˆ›å»ºgseaåˆ†æçš„geneListï¼ˆåŒ…å«ä»å¤§åˆ°å°æ’åˆ—çš„log2FoldChangeå’ŒENTREZIDä¿¡æ¯ï¼‰
CS3_DEG_trans <- merge(CS3_DEG_trans, SYMBOL_ENTERZID , by='SYMBOL') #æŒ‰ç…§SYMBOLåˆå¹¶æ³¨é‡Šä¿¡æ¯
CS3_geneList <- CS3_DEG_trans$log2fc
names(CS3_geneList) <- CS3_DEG_trans$ENTREZID
CS3_geneList <- sort(CS3_geneList, decreasing = T)  #ä»å¤§åˆ°å°æ’åº

##gseaå¯Œé›† 
CS3_KEGG_kk_entrez <- gseKEGG(geneList     = CS3_geneList,
                              organism     = organism, #äººhsa é¼ mmu
                              pvalueCutoff = 0.25)  #å®é™…ä¸ºpadjé˜ˆå€¼å¯è°ƒæ•´ 
CS3_KEGG_kk <- DOSE::setReadable(CS3_KEGG_kk_entrez, 
                                 OrgDb=OrgDb,
                                 keyType='ENTREZID')#è½¬åŒ–id             

CS3_GO_kk_entrez <- gseGO(geneList     = CS3_geneList,
                          ont          = "ALL",  # "BP"ã€"MF"å’Œ"CC"æˆ–"ALL"
                          OrgDb        = OrgDb,
                          keyType      = "ENTREZID",
                          pvalueCutoff = 0.25)   #å®é™…ä¸ºpadjé˜ˆå€¼å¯è°ƒæ•´
CS3_GO_kk <- DOSE::setReadable(CS3_GO_kk_entrez, 
                               OrgDb=OrgDb,
                               keyType='ENTREZID')#è½¬åŒ–id 

save(CS3_KEGG_kk_entrez, CS3_GO_kk_entrez, file = "CS3_GSEA_result.RData")

#KEGGç»“æœå¯è§†åŒ–
##é€‰å–å¯Œé›†ç»“æœ
CS3_kk_gse <- CS3_KEGG_kk
###æ¡ä»¶ç­›é€‰
CS3_kk_gse_cut <- CS3_kk_gse[CS3_kk_gse$pvalue<0.05 & CS3_kk_gse$p.adjust<0.25 & abs(CS3_kk_gse$NES)>1]
#é€‰æ‹©å±•ç°NESå‰å‡ ä¸ªé€šè·¯ 
CS3_diff_gsea <- CS3_kk_gse_cut[head(order(abs(CS3_kk_gse_cut$NES),decreasing = T),10),]
#### åˆå¹¶ GSEAé€šè·¯ 
CS3_gseap <- gseaplot2(CS3_kk_gse,
                       CS3_diff_gsea$ID,#å¯Œé›†çš„IDç¼–å·
                       title = "CS3 KEGG Pathway",#æ ‡é¢˜
                       color = "red",#GSEAçº¿æ¡é¢œè‰²
                       base_size = 20,#åŸºç¡€å­—ä½“å¤§å°
                       rel_heights = c(1.5, 0.5, 1),#å‰¯å›¾çš„ç›¸å¯¹é«˜åº¦
                       subplots = 1:3, #è¦æ˜¾ç¤ºå“ªäº›å‰¯å›¾ å¦‚subplots=c(1,3) #åªè¦ç¬¬ä¸€å’Œç¬¬ä¸‰ä¸ªå›¾
                       ES_geom = "line",#enrichment scoreç”¨çº¿è¿˜æ˜¯ç”¨ç‚¹"dot"
                       pvalue_table = F) #æ˜¾ç¤ºpvalueç­‰ä¿¡æ¯
ggsave(CS3_gseap, filename = "CS3_GSEA_KEGG.png",width =16,height =16,dpi = 600)

#GOç»“æœå¯è§†åŒ–
# ç»˜åˆ¶ç‚¹å›¾ï¼ˆæŒ‰p.adjustæ’åºï¼Œåˆ†é¢æ˜¾ç¤ºGOç±»åˆ«ï¼‰
dotplot(CS3_GO_kk,
        x = "NES",           # Xè½´ä¸ºNESå€¼ï¼ˆåæ˜ å¯Œé›†æ–¹å‘ï¼‰
        color = "p.adjust",  # é¢œè‰²è¡¨ç¤ºp.adjust
        showCategory = 10,   # æ¯ç±»æ˜¾ç¤ºTop10é€šè·¯
        split = "ONTOLOGY",  # æŒ‰GOç±»åˆ«åˆ†é¢
        font.size = 8) +  
  facet_grid(ONTOLOGY ~ ., scales = "free_y", space = "free") +
  ggtitle("CS3 GO Enrichment Analysis") +
  scale_color_gradient(
    low = "red", high = "blue",
    limits = c(0, 1))+
  
  theme(plot.title = element_text(hjust = 0.5))  
ggsave("CS3_GO_Dotplot.png", width = 8, height = 12,dpi = 600) 




#4ã€CS4äºšå‹å¯Œé›†åˆ†æ
#æå–äºš4 vs å…¶ä»–äºšå‹çš„log2FCæ’åºåŸºå› åˆ—è¡¨

CS4_DEG_trans <- CS4_vs_others[,2,drop=FALSE]
CS4_DEG_trans$SYMBOL <- rownames(CS4_DEG_trans)

##åˆ›å»ºgseaåˆ†æçš„geneListï¼ˆåŒ…å«ä»å¤§åˆ°å°æ’åˆ—çš„log2FoldChangeå’ŒENTREZIDä¿¡æ¯ï¼‰
CS4_DEG_trans <- merge(CS4_DEG_trans, SYMBOL_ENTERZID , by='SYMBOL') #æŒ‰ç…§SYMBOLåˆå¹¶æ³¨é‡Šä¿¡æ¯
CS4_geneList <- CS4_DEG_trans$log2fc
names(CS4_geneList) <- CS4_DEG_trans$ENTREZID
CS4_geneList <- sort(CS4_geneList, decreasing = T)  #ä»å¤§åˆ°å°æ’åº

##gseaå¯Œé›† 
CS4_KEGG_kk_entrez <- gseKEGG(geneList     = CS4_geneList,
                              organism     = organism, #äººhsa é¼ mmu
                              pvalueCutoff = 0.25)  #å®é™…ä¸ºpadjé˜ˆå€¼å¯è°ƒæ•´ 
CS4_KEGG_kk <- DOSE::setReadable(CS4_KEGG_kk_entrez, 
                                 OrgDb=OrgDb,
                                 keyType='ENTREZID')#è½¬åŒ–id             

CS4_GO_kk_entrez <- gseGO(geneList     = CS4_geneList,
                          ont          = "ALL",  # "BP"ã€"MF"å’Œ"CC"æˆ–"ALL"
                          OrgDb        = OrgDb,
                          keyType      = "ENTREZID",
                          pvalueCutoff = 0.25)   #å®é™…ä¸ºpadjé˜ˆå€¼å¯è°ƒæ•´
CS4_GO_kk <- DOSE::setReadable(CS4_GO_kk_entrez, 
                               OrgDb=OrgDb,
                               keyType='ENTREZID')#è½¬åŒ–id 

save(CS4_KEGG_kk_entrez, CS4_GO_kk_entrez, file = "CS4_GSEA_result.RData")

#KEGGç»“æœå¯è§†åŒ–
##é€‰å–å¯Œé›†ç»“æœ
CS4_kk_gse <- CS4_KEGG_kk
###æ¡ä»¶ç­›é€‰
CS4_kk_gse_cut <- CS4_kk_gse[CS4_kk_gse$pvalue<0.05 & CS4_kk_gse$p.adjust<0.25 & abs(CS4_kk_gse$NES)>1]
#é€‰æ‹©å±•ç°NESå‰å‡ ä¸ªé€šè·¯ 
CS4_diff_gsea <- CS4_kk_gse_cut[head(order(abs(CS4_kk_gse_cut$NES),decreasing = T),10),]
#### åˆå¹¶ GSEAé€šè·¯ 
CS4_gseap <- gseaplot2(CS4_kk_gse,
                       CS4_diff_gsea$ID,#å¯Œé›†çš„IDç¼–å·
                       title = "CS4 KEGG Pathway",#æ ‡é¢˜
                       color = "red",#GSEAçº¿æ¡é¢œè‰²
                       base_size = 20,#åŸºç¡€å­—ä½“å¤§å°
                       rel_heights = c(1.5, 0.5, 1),#å‰¯å›¾çš„ç›¸å¯¹é«˜åº¦
                       subplots = 1:3, #è¦æ˜¾ç¤ºå“ªäº›å‰¯å›¾ å¦‚subplots=c(1,3) #åªè¦ç¬¬ä¸€å’Œç¬¬ä¸‰ä¸ªå›¾
                       ES_geom = "line",#enrichment scoreç”¨çº¿è¿˜æ˜¯ç”¨ç‚¹"dot"
                       pvalue_table = F) #æ˜¾ç¤ºpvalueç­‰ä¿¡æ¯
ggsave(CS4_gseap, filename = "CS4_GSEA_KEGG.png",width =16,height =16,dpi = 600)

#GOç»“æœå¯è§†åŒ–
# ç»˜åˆ¶ç‚¹å›¾ï¼ˆæŒ‰p.adjustæ’åºï¼Œåˆ†é¢æ˜¾ç¤ºGOç±»åˆ«ï¼‰
dotplot(CS4_GO_kk,
        x = "NES",           # Xè½´ä¸ºNESå€¼ï¼ˆåæ˜ å¯Œé›†æ–¹å‘ï¼‰
        color = "p.adjust",  # é¢œè‰²è¡¨ç¤ºp.adjust
        showCategory = 10,   # æ¯ç±»æ˜¾ç¤ºTop10é€šè·¯
        split = "ONTOLOGY",  # æŒ‰GOç±»åˆ«åˆ†é¢
        font.size = 8) +  
  facet_grid(ONTOLOGY ~ ., scales = "free_y", space = "free") +
  ggtitle("CS4 GO Enrichment Analysis") +
  scale_color_gradient(low = "red", high = "blue") +
  theme(plot.title = element_text(hjust = 0.5))  # è‡ªå®šä¹‰é¢œè‰²
ggsave("CS4_GO_Dotplot.png", width = 8, height = 12,dpi = 600) 


#5ã€CS5äºšå‹å¯Œé›†åˆ†æ
#æå–äºšå‹5vs å…¶ä»–äºšå‹çš„log2FCæ’åºåŸºå› åˆ—è¡¨
CS5_DEG_trans <- CS5_vs_others[,2,drop=FALSE]
CS5_DEG_trans$SYMBOL <- rownames(CS5_DEG_trans)

##åˆ›å»ºgseaåˆ†æçš„geneListï¼ˆåŒ…å«ä»å¤§åˆ°å°æ’åˆ—çš„log2FoldChangeå’ŒENTREZIDä¿¡æ¯ï¼‰
CS5_DEG_trans <- merge(CS5_DEG_trans, SYMBOL_ENTERZID , by='SYMBOL') #æŒ‰ç…§SYMBOLåˆå¹¶æ³¨é‡Šä¿¡æ¯
CS5_geneList <- CS5_DEG_trans$log2fc
names(CS5_geneList) <- CS5_DEG_trans$ENTREZID
CS5_geneList <- sort(CS5_geneList, decreasing = T)  #ä»å¤§åˆ°å°æ’åº

##gseaå¯Œé›† 
CS5_KEGG_kk_entrez <- gseKEGG(geneList     = CS5_geneList,
                              organism     = organism, #äººhsa é¼ mmu
                              pvalueCutoff = 0.25)  #å®é™…ä¸ºpadjé˜ˆå€¼å¯è°ƒæ•´ 
CS5_KEGG_kk <- DOSE::setReadable(CS5_KEGG_kk_entrez, 
                                 OrgDb=OrgDb,
                                 keyType='ENTREZID')#è½¬åŒ–id             

CS5_GO_kk_entrez <- gseGO(geneList     = CS5_geneList,
                          ont          = "ALL",  # "BP"ã€"MF"å’Œ"CC"æˆ–"ALL"
                          OrgDb        = OrgDb,
                          keyType      = "ENTREZID",
                          pvalueCutoff = 0.25)   #å®é™…ä¸ºpadjé˜ˆå€¼å¯è°ƒæ•´
CS5_GO_kk <- DOSE::setReadable(CS5_GO_kk_entrez, 
                               OrgDb=OrgDb,
                               keyType='ENTREZID')#è½¬åŒ–id 

save(CS5_KEGG_kk_entrez, CS5_GO_kk_entrez, file = "CS5_GSEA_result.RData")

#KEGGç»“æœå¯è§†åŒ–
##é€‰å–å¯Œé›†ç»“æœ
CS5_kk_gse <- CS5_KEGG_kk
###æ¡ä»¶ç­›é€‰
CS5_kk_gse_cut <- CS5_kk_gse[CS5_kk_gse$pvalue<0.05 & CS5_kk_gse$p.adjust<0.25 & abs(CS5_kk_gse$NES)>1]
#é€‰æ‹©å±•ç°NESå‰å‡ ä¸ªé€šè·¯ 
CS5_diff_gsea <- CS5_kk_gse_cut[head(order(abs(CS5_kk_gse_cut$NES),decreasing = T),10),]
#### åˆå¹¶ GSEAé€šè·¯ 
CS5_gseap <- gseaplot2(CS5_kk_gse,
                       CS5_diff_gsea$ID,#å¯Œé›†çš„IDç¼–å·
                       title = "CS5 KEGG Pathway",#æ ‡é¢˜
                       color = "red",#GSEAçº¿æ¡é¢œè‰²
                       base_size = 20,#åŸºç¡€å­—ä½“å¤§å°
                       rel_heights = c(1.5, 0.5, 1),#å‰¯å›¾çš„ç›¸å¯¹é«˜åº¦
                       subplots = 1:3, #è¦æ˜¾ç¤ºå“ªäº›å‰¯å›¾ å¦‚subplots=c(1,3) #åªè¦ç¬¬ä¸€å’Œç¬¬ä¸‰ä¸ªå›¾
                       ES_geom = "line",#enrichment scoreç”¨çº¿è¿˜æ˜¯ç”¨ç‚¹"dot"
                       pvalue_table = F) #æ˜¾ç¤ºpvalueç­‰ä¿¡æ¯
ggsave(CS5_gseap, filename = "CS5_GSEA_KEGG.png",width =16,height =16,dpi = 600)

#GOç»“æœå¯è§†åŒ–
# ç»˜åˆ¶ç‚¹å›¾ï¼ˆæŒ‰p.adjustæ’åºï¼Œåˆ†é¢æ˜¾ç¤ºGOç±»åˆ«ï¼‰
dotplot(CS5_GO_kk,
        x = "NES",           # Xè½´ä¸ºNESå€¼ï¼ˆåæ˜ å¯Œé›†æ–¹å‘ï¼‰
        color = "p.adjust",  # é¢œè‰²è¡¨ç¤ºp.adjust
        showCategory = 10,   # æ¯ç±»æ˜¾ç¤ºTop10é€šè·¯
        split = "ONTOLOGY",  # æŒ‰GOç±»åˆ«åˆ†é¢
        font.size = 8) +  
  facet_grid(ONTOLOGY ~ ., scales = "free_y", space = "free") +
  ggtitle("CS5 GO Enrichment Analysis") +
  scale_color_gradient(low = "red", high = "blue") +
  theme(plot.title = element_text(hjust = 0.5))  # è‡ªå®šä¹‰é¢œè‰²
ggsave("CS5_GO_Dotplot.png", width = 8, height = 12,dpi = 600) 





# åˆå¹¶æ‰€æœ‰äºšå‹çš„GOå¯Œé›†ç»“æœ
combined_GO <- rbind(
  cbind(CS1_GO_kk@result, Subtype = "CS1"),
  cbind(CS2_GO_kk@result, Subtype = "CS2"),
  cbind(CS3_GO_kk@result, Subtype = "CS3"),
  cbind(CS4_GO_kk@result, Subtype = "CS4"),
  cbind(CS5_GO_kk@result, Subtype = "CS5")
)

# æŒ‰äºšå‹å’ŒGOç±»åˆ«ç­›é€‰Top10é€šè·¯ï¼ˆæŒ‰p.adjustæ’åºï¼‰
library(dplyr)
combined_GO_top10 <- combined_GO %>%
  group_by(Subtype, ONTOLOGY) %>%
  arrange(p.adjust, .by_group = TRUE) %>%
  slice_head(n = 10) %>%
  ungroup()
# ğŸ› ï¸ å…³é”®ä¼˜åŒ–æ­¥éª¤ï¼šæ·»åŠ æ–‡æœ¬æ¢è¡Œå¤„ç†
library(stringr)
combined_GO_top10 <- combined_GO_top10 %>% 
  mutate(
    Description = str_wrap(Description, 
                           width = 70,  # æ¯è¡Œæœ€å¤š40å­—ç¬¦
                           exdent = 2)  # ç¬¬äºŒè¡Œç¼©è¿›2å­—ç¬¦
  )
# ç¡®ä¿äºšå‹é¡ºåºæ­£ç¡®
combined_GO_top10$Subtype <- factor(combined_GO_top10$Subtype, 
                                 levels = c("CS1", "CS2", "CS3", "CS4", "CS5"))
# åˆ›å»ºåˆ†é¢ç‚¹å›¾
library(ggplot2)
GO_plot1 <- ggplot(combined_GO_top10, 
            aes(x = NES, 
                y = reorder(Description, NES),  # æŒ‰NESæ’åºyè½´
                color = p.adjust,
                size = setSize)) +
  geom_point(alpha = 0.8) +
  scale_color_gradient(low = "blue", high = "red",  # é¢œè‰²æ˜ å°„p.adjust
                       limits = c(0, 0.003),         # ç»Ÿä¸€é¢œè‰²æ ‡å°º
                       name = "Adjusted p-value") +
  scale_size_continuous(name = "Gene Count",        # ç‚¹å¤§å°æ˜ å°„åŸºå› æ•°é‡
                        range = c(2, 6)) +
  facet_grid(ONTOLOGY ~ Subtype,                    # çŸ©é˜µåˆ†é¢
             scales = "free_y",                     # è‡ªç”±yè½´èŒƒå›´
             space = "free") +                      # è‡ªåŠ¨è°ƒæ•´é—´è·
  labs(x = "Normalized Enrichment Score (NES)",
       y = "GO Pathway",
       title = "Multi-Subtype GO Enrichment Analysis") +
  theme_bw(base_size = 12) +
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold", size = 14),
    axis.text.y = element_text(size = 8),
    axis.text.x = element_text(angle = 45, hjust = 1),
    strip.text = element_text(face = "bold"),
    panel.spacing = unit(0.5, "lines")
  )
GO_plot1
# ä¿å­˜é«˜æ¸…å›¾ï¼ˆå¯æ ¹æ®éœ€è¦è°ƒæ•´å°ºå¯¸ï¼‰
ggsave("Combined_GO_Dotplot.pdf", GO_plot1,   width = 20, height = 15, dpi = 600)





#å¤šç»„ç»“æœå¯¹æ¯”â€”â€”KEGG
allsubtype_genelist <- list(CS1_geneList,CS2_geneList,CS3_geneList,CS4_geneList,CS5_geneList)

lapply(1:5, function(x){
  CS_kegg <- gseKEGG(geneList     = allsubtype_genelist[[x]],
                organism = "hsa",
                keyType = "kegg",
                minGSSize    = 100,
                maxGSSize    = 500,
                pvalueCutoff = 1,#è®¾ç½® pvalueCutoff=1,ä¿è¯åœ¨æ¯ä¸ªå¯Œé›†ç»“æœé‡Œéƒ½æœ‰å…±åŒçš„é€šè·¯
                eps          = 0,  #ç²¾ç¡®è®¡ç®—æå°på€¼
                verbose      = FALSE)
}) -> allsubtypes_gsea_kegg_list
#æŸ¥çœ‹å…¶ä¸­ä¸€ä¸ªç»“æœ
View(as.data.frame(allsubtypes_gsea_kegg_list[[1]]))
View(as.data.frame(allsubtypes_gsea_kegg_list[[4]]))
GSEAmultiGP(gsea_list = allsubtypes_gsea_kegg_list,
            geneSetID = "hsa04110",#cell type
            exp_name = c("CS1","CS2","CS3","CS4","CS5"),
            addPval = T,
            pvalX = 0.99,pvalY = 0.99,
            legend.position = "right",
            curve.col = ggsci::pal_lancet()(5))

ggsave("allsubtypes_gsea_kegg1.png",width = 8,height = 8,dpi=900)

GSEAmultiGP(gsea_list = allsubtypes_gsea_kegg_list,
            geneSetID = "hsa04060", #Cytokine-cytokine receptor interaction
            exp_name = c("CS1","CS2","CS3","CS4","CS5"),
            addPval = T,
            pvalX = 0.99,pvalY = 0.99,
            legend.position = "right",
            curve.col = ggsci::pal_lancet()(5))

ggsave("allsubtypes_gsea_kegg2.png",width = 8,height = 8,dpi=900)


GSEAmultiGP(gsea_list = allsubtypes_gsea_kegg_list,
            geneSetID = "hsa04659", #Th17 cell differentiation
            exp_name = c("CS1","CS2","CS3","CS4","CS5"),
            addPval = T,
            pvalX = 0.99,pvalY = 0.99,
            legend.position = "right",
            curve.col = ggsci::pal_lancet()(5))

ggsave("allsubtypes_gsea_kegg3.png",width = 8,height = 8,dpi=900)

GSEAmultiGP(gsea_list = allsubtypes_gsea_kegg_list,
            geneSetID = "hsa04145", #Phagosome
            exp_name = c("CS1","CS2","CS3","CS4","CS5"),
            addPval = T,
            pvalX = 0.99,pvalY = 0.99,
            legend.position = "right",
            curve.col = ggsci::pal_lancet()(5))

ggsave("allsubtypes_gsea_kegg4.png",width = 8,height = 8,dpi=900)

GSEAmultiGP(gsea_list = allsubtypes_gsea_kegg_list,
            geneSetID = "hsa04062", #Chemokine signaling pathway
            exp_name = c("CS1","CS2","CS3","CS4","CS5"),
            addPval = T,
            pvalX = 0.99,pvalY = 0.99,
            legend.position = "right",
            curve.col = ggsci::pal_lancet()(5))

ggsave("allsubtypes_gsea_kegg5.png",width = 8,height = 8,dpi=900)

GSEAmultiGP(gsea_list = allsubtypes_gsea_kegg_list,
            geneSetID = "hsa04670", #Leukocyte transendothelial migration
            exp_name = c("CS1","CS2","CS3","CS4","CS5"),
            addPval = T,
            pvalX = 0.99,pvalY = 0.99,
            legend.position = "right",
            curve.col = ggsci::pal_lancet()(5))
ggsave("allsubtypes_gsea_kegg6.png",width = 8,height = 8,dpi=900)


#å¤šç»„ç»“æœå¯¹æ¯”â€”â€”GO
lapply(1:5, function(x){
  CS_go <- gseGO(geneList     = allsubtype_genelist[[x]],
                 OrgDb = org.Hs.eg.db,
                 keyType      = "ENTREZID",
                 ont = "BP",
                 minGSSize    = 100,
                 maxGSSize    = 500,
                pvalueCutoff = 1,#è®¾ç½® pvalueCutoff=1,ä¿è¯åœ¨æ¯ä¸ªå¯Œé›†ç»“æœé‡Œéƒ½æœ‰å…±åŒçš„é€šè·¯
                eps          = 0,  #ç²¾ç¡®è®¡ç®—æå°på€¼
                verbose      = FALSE)
}) -> allsubtypes_gsea_go_list
#æŸ¥çœ‹å…¶ä¸­ä¸€ä¸ªç»“æœ
View(as.data.frame(allsubtypes_gsea_go_list[[4]]))
#é€‰æ‹©æ„Ÿå…´è¶£çš„GOæ¡ç›®ç»˜å›¾
GSEAmultiGP(gsea_list = allsubtypes_gsea_go_list,
            geneSetID = "GO:0002443",#leukocyte mediated immunity
            exp_name = c("CS1","CS2","CS3","CS4","CS5"),
            addPval = T,
            pvalX = 0.99,pvalY = 0.99,
            legend.position = "right",
            curve.col = ggsci::pal_lancet()(5))
ggsave("allsubtypes_gsea_go1.png",width = 8,height = 8,dpi=900)

GSEAmultiGP(gsea_list = allsubtypes_gsea_go_list,
            geneSetID = "GO:0002449",#lymphocyte mediated immunity
            exp_name = c("CS1","CS2","CS3","CS4","CS5"),
            addPval = T,
            pvalX = 0.99,pvalY = 0.99,
            legend.position = "right",
            curve.col = ggsci::pal_lancet()(5))
ggsave("allsubtypes_gsea_go2.png",width = 8,height = 8,dpi=900)

GSEAmultiGP(gsea_list = allsubtypes_gsea_go_list,
            geneSetID = "GO:0002274",#myeloid leukocyte activation
            exp_name = c("CS1","CS2","CS3","CS4","CS5"),
            addPval = T,
            pvalX = 0.99,pvalY = 0.99,
            legend.position = "right",
            curve.col = ggsci::pal_lancet()(5))
ggsave("allsubtypes_gsea_go3.png",width = 8,height = 8,dpi=900)


GSEAmultiGP(gsea_list = allsubtypes_gsea_go_list,
            geneSetID = "GO:0002377",#immunoglobulin production
            exp_name = c("CS1","CS2","CS3","CS4","CS5"),
            addPval = T,
            pvalX = 0.99,pvalY = 0.99,
            legend.position = "right",
            curve.col = ggsci::pal_lancet()(5))
ggsave("allsubtypes_gsea_go4.png",width = 8,height = 8,dpi=900)


GSEAmultiGP(gsea_list = allsubtypes_gsea_go_list,
            geneSetID = "GO:0019724",#B cell mediated immunity
            exp_name = c("CS1","CS2","CS3","CS4","CS5"),
            addPval = T,
            pvalX = 0.99,pvalY = 0.99,
            legend.position = "right",
            curve.col = ggsci::pal_lancet()(5))
ggsave("allsubtypes_gsea_go5.png",width = 8,height = 8,dpi=900)


GSEAmultiGP(gsea_list = allsubtypes_gsea_go_list,
            geneSetID = "GO:0002460",#adaptive immune response based on somatic recombination of immune receptors built from immunoglobulin superfamily domains

            exp_name = c("CS1","CS2","CS3","CS4","CS5"),
            addPval = T,
            pvalX = 0.99,pvalY = 0.99,
            legend.position = "right",
            curve.col = ggsci::pal_lancet()(5))
ggsave("allsubtypes_gsea_go6.png",width = 8,height = 8,dpi=900)


setwd("~/NSCLC_subtyping/Downstream Analysis of MOVICS clusters")
save.image("Downstream Analysis of MOVICS clusters.RData")
